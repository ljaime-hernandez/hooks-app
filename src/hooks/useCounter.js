import {useState} from 'react';

export const useCounter = (initialState = 10) => {
    
    const [state, setState] = useState(initialState);

    const increment = (factor = 1) => {
        setState(state + factor);
    }

    const decrement = (factor = 1) => {
        setState(state - factor);
    }

    const reset = () => {
        setState(initialState);
    }

    return {
        state,
        increment,
        decrement,
        reset
    };
}
