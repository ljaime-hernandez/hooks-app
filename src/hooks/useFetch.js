import { useEffect, useRef, useState } from 'react'

export const useFetch = (url) => {

    // with the help of the useRef, we are avoiding an error displaying on the console
    // which will be triggered if we mount and then dismount the component when it has
    // not finalized the processes on it, in this case the fetch, the useRef then is 
    // used so we can track the launch of this hook in the component, refer to the first
    // useEffect hook for more information
    const isMounted = useRef(true);

    // we add default values for the component using this hook to render accordingly, 
    // when the data have not been retrieved from the API, then its declared as null,
    // a loading value is set to true so we can render certain elements on the component
    // for both visual effect and for the user to know if the component is working or not,
    // the error value is created just in case theres an error loading the data.
    const [state, setState] = useState({data: null, loading: true, error: null});

    // this useEffect does not has an input (empty brackets at the end), therefore it will
    // listen to any action taken in this hook, for this purpose, we set the default value of the
    // useRef to false so the fetch in this hook does not do any action when the component is
    // dismounted, avoiding the web for crashing.
    useEffect(() => {
        return () => {
            isMounted.current = false
        }
    }, [])

    // the useEffect is going to load whenever a change to the url is made, in this case,
    // the url is modified from the MultipleCustomHooks component with the help of a button
    // (which is also a hook)
    useEffect(() => {

        // simple state reset for the loading to return to true after being previously used
        setState({data: null, loading: true, error: null});
        
        // the fetch will use the url to retrieve data from our API, that information is then
        // transformed into JSON format, and finally we use the setState to both change the values
        // of the state and to add the data received from the fetch request 
        fetch(url)
            .then(res => res.json())
            .then(data => {

                // the useRef will allow the hook to work only when the component
                // is rendered, if it is dismounted then this action will not be taken
                if(isMounted.current){

                setTimeout(() => {
                    setState({
                        loading: false,
                        error: null,
                        data
                })}, 2000)}
            }).catch(() => {
                setState({
                    data: null,
                    loading: false,
                    error: 'Info could not get loaded'
                })
            })
    }, [url]);

    return state;
}