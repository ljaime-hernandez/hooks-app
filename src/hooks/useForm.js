import { useState } from 'react';

// the useForm hook will receive a value but will have an empty object as its initial
// state, we create a function called handleInputChange which will return a value based 
// on the name of the input where we placed it, means that 
export const useForm = (initialState = {}) => {

    const [values, setValues] = useState(initialState);

    const reset = () => {
        setValues(initialState);
    }
    // the handleInputChange function will receive the values from the FormWithCustomHook 
    // component, the spread operator is used so the variables on it, either empty or not, 
    // remain the same while we receive the event (as this function is used in the 'onChange'
    // event) variable, as we already know the event will contain a target value, we destructure
    // it for easier readability, then we use the name of the input previously written on the
    // form and use it in our advantage to actually change the value on it, allowing us to
    // modify the value on the object from the FormWithCustomHook object as well, as it is passed
    // by reference
    const handleInputChange = ({target}) => {
        setValues({
            ...values,
            [target.name]: target.value
        });
    }
    return [values, handleInputChange, reset];
}
