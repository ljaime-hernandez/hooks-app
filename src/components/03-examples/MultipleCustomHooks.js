import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';
import '../02-useEffect/effects.css';

export const MultipleCustomHooks = () => {

    // custom hook used for other components but using a different number as argument, in this
    // component we will need just both the value and the increment function from it. The increment
    // function is used on the button rendered on this component in a click event which will increase the
    // counter by one, changing the state on the useFetch, which will therefore make a new fetch request
    const {state: counter, increment} = useCounter(1);
    // the counter in the url is modified with a button, which will be used as the id required
    // for use to retrieve the data from the page
    const {loading, data} = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    // the following syntax is used so we dont have an 'undefined' data to be checked before
    // we render its information into the component, but its going to confirm if either the data exist
    // or not, the operator !! will return either a true or a false. Also, similar to a ternary operation,
    // depending if the condition is true, then we will destructure both the quote and the author
    // (previously confirmed with the help of Postman) retrieved from the first index of an array, 
    // as the information returned is an object contained on an array, if confirmed with a console.log,
    // the web console displays a [{...}], pushing us to retrieve the information of the data array by index.
    const {author, quote} = !!data && data[0];

    return (
        <div>
            <h1>Multiple custom hooks demo (Breaking Bad Quotes API)</h1>
            {/*ternary operation used with the help of the loading value retrieved from the useFetch hook,
                if the information dont exist, it will display a loading element as an alert, else if the data
                exists, it will display a blockquote with certain classes for better display */}
            {
                loading
                ?
                    (
                        <div className='alert alert-info text-center'>
                            Loading...
                        </div>
                    )
                :
                    (
                        <blockquote className="blockquote text-end">
                            <p className="quote">{quote}</p>
                            <figcaption className="blockquote-footer">{author}</figcaption>
                        </blockquote>
                    )
            }

            <button className='btn btn-primary' onClick={increment}>
                Change Quote
            </button>
        </div>
    )
}
