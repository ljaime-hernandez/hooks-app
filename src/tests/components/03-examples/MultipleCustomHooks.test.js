import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks"
import { useCounter } from "../../../hooks/useCounter";
import { useFetch } from "../../../hooks/useFetch";

// two of the custom hooks are used on this component, thats why we need to use the
// jest.mock method in both of them to import the return object which each of them will
// bring to us
jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");

describe('tests in MultipleCustomHooks component', () => {


    beforeEach(()=>{
        useCounter.mockReturnValue({
            state: 1,
            increment: () => {}
        })
    })

    test('should render properly', () => {
        
        // default values on the useFetch hook which should actually
        // allow for a div component to render if the data is null and the loading
        // state is true
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks/>);
        expect(wrapper).toMatchSnapshot();
    })

    test('should show information', () => {
        
        // In this case we play as if the fetch returned an object with data, 
        // if seen on the actual custom hook, the loading should turn false and the
        // data should give us both an author and a quote
        useFetch.mockReturnValue({
            data: [{
                author: 'Miguel',
                quote: 'Hello World'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks/>);

        // this expect will confirm if the loading state is false

        expect(wrapper.find('.alert').exists()).toBe(false);
        // the text inside the element with a className of quote should return the
        // quote
        expect(wrapper.find('.quote').text().trim()).toBe('Hello World');
        // the text inside the element with a className of blockquote-footer should return the
        // authors name
        expect(wrapper.find('.blockquote-footer').text().trim()).toBe('Miguel');
    })
})