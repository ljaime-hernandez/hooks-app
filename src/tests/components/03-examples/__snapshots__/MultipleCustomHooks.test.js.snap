// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in MultipleCustomHooks component should render properly 1`] = `
<div>
  <h1>
    Multiple custom hooks demo (Breaking Bad Quotes API)
  </h1>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Change Quote
  </button>
</div>
`;

exports[`tests in MultipleCustomHooks should render properly 1`] = `
<div>
  <h1>
    Multiple custom hooks demo (Breaking Bad Quotes API)
  </h1>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Change Quote
  </button>
</div>
`;
